### PR TITLE
Add MIG support for H100 GPU

### PR DESCRIPTION
**Pull request by Jiaqicao257**
Manually tested with A100 40GB and partition size 1g.5gb and A100 80GB and H100 80GB with partition size 1g.10gb

### CODE REVIEW

File: partition_gpu/partition_gpu.go [hunk 1]
```diff
@@ -98,18 +107,27 @@ func main() {
...
-		}
+		if gpuType == Nvidia40gbA100 || gpuType == Nvidia80gbA100 {
+			glog.Infof("Rebooting node to enable MIG mode")
+			if err := rebootNode(); err != nil {
+				glog.Errorf("Failed to trigger node reboot after enabling MIG mode: %v", err)
+			}
 
-		// Exit, since we cannot proceed until node has rebooted, for MIG changes to take effect.
-		os.Exit(1)
+			// Exit, since we cannot proceed until node has rebooted, for MIG changes to take effect.
```
Comment (shakeebshams) [inline partition_gpu/partition_gpu.go:128]: given the comment, might want to exit anyway is mig mode isnt enabled. So there should should be a a os.Exit(1) outside of this `if gpuType` clause

File: partition_gpu/partition_gpu.go [hunk 2]
```diff
@@ -98,18 +107,27 @@ func main() {
...
 	}
 	if !migModeEnabled {
+		glog.Infof("Checking the GPU type now.")
+		gpuType, err := checkGpuType()
+		if err != nil {
+			glog.Errorf("Failed to check GPU Type: %v", err)
+			os.Exit(1)
+		}
+		glog.Infof("Got GPU type used: %s", gpuType)
 		glog.Infof("MIG mode is not enabled. Enabling now.")
```
Comment (grac3gao) [inline partition_gpu/partition_gpu.go:117]: Move `glog.Infof("MIG mode is not enabled. Enabling now.")` before  `glog.Infof("Checking the GPU type now.")`

File: partition_gpu/partition_gpu.go [hunk 3]
```diff
@@ -98,18 +107,27 @@ func main() {
...
 		if err := enableMigMode(); err != nil {
 			glog.Errorf("Failed to enable MIG mode: %v", err)
 			os.Exit(1)
 		}
-		glog.Infof("Rebooting node to enable MIG mode")
-		if err := rebootNode(); err != nil {
-			glog.Errorf("Failed to trigger node reboot after enabling MIG mode: %v", err)
-		}
+		if gpuType == Nvidia40gbA100 || gpuType == Nvidia80gbA100 {
```
Comment (grac3gao) [inline partition_gpu/partition_gpu.go:122]: Add more comments for why a100 can reboot, whereas h100 doesn't reboot. And add reference: https://docs.nvidia.com/datacenter/tesla/mig-user-guide/#reset-hopper

File: partition_gpu/partition_gpu.go (hunk 1 above)
Comment (grac3gao) [reply partition_gpu/partition_gpu.go:128]: os.Exit(1) is only needed for Nvidia40gbA100 and Nvidia80gbA100, @Jiaqicao257 please add more comments to explain this logic.

File: partition_gpu/partition_gpu.go [hunk 4]
```diff
@@ -45,6 +45,8 @@ var partitionSizeToProfileID = map[string]string{
 	"3g.40gb": "9",
 	"4g.40gb": "5",
 	"7g.80gb": "0",
+	//nvidia-h100-80gb
+	"1g.20gb": "15",
```
Comment (grac3gao) [inline partition_gpu/partition_gpu.go:49]: The comment is mis-leading. h100 actually has all the same profiles as nvidia-a100-80gb + "1g.20gb", please reflect this.

File: pkg/gpu/nvidia/mig/mig.go [hunk 5]
```diff
@@ -41,6 +41,8 @@ var gpuPartitionSizeMaxCount = map[string]int{
 	"2g.20gb": 3,
 	"3g.40gb": 2,
 	"7g.80gb": 1,
+	//nvidia-h100-80gb
```
Comment (grac3gao) [inline pkg/gpu/nvidia/mig/mig.go:44]: Ditto, please update the comments.

**Comment by grac3gao** [comment]
@Jiaqicao257 Do we manually validate this works?

File: partition_gpu/partition_gpu.go [hunk 6]
```diff
@@ -169,6 +187,24 @@ func enableMigMode() error {
 	return exec.Command(*nvidiaSmiPath, "-mig", "1").Run()
 }
 
+// checkGpuType checkes the GPU type used
+func checkGpuType() (string, error) {
+	gpuType, err := exec.Command(*nvidiaSmiPath, "--query-gpu=gpu_name", "--format=csv,noheader").Output()
+	if err != nil {
+		return "", err
+	}
+	if strings.HasPrefix(string(gpuType), Nvidia80gbH100) {
```
Comment (grac3gao) [inline partition_gpu/partition_gpu.go:196]: Can we use switch here instead of if..else...?

File: partition_gpu/partition_gpu.go (hunk 3 above)
Comment (Jiaqicao257) [reply partition_gpu/partition_gpu.go:125]: Done, added references

File: pkg/gpu/nvidia/mig/mig.go (hunk 5 above)
Comment (Jiaqicao257) [reply pkg/gpu/nvidia/mig/mig.go:44]: Done.

File: partition_gpu/partition_gpu.go (hunk 4 above)
Comment (Jiaqicao257) [reply partition_gpu/partition_gpu.go:49]: Done.

File: partition_gpu/partition_gpu.go (hunk 1 above)
Comment (Jiaqicao257) [reply partition_gpu/partition_gpu.go]: Added comments to clarify this

File: partition_gpu/partition_gpu.go (hunk 2 above)
Comment (Jiaqicao257) [reply partition_gpu/partition_gpu.go]: Done

File: partition_gpu/partition_gpu.go (hunk 6 above)
Comment (Jiaqicao257) [reply partition_gpu/partition_gpu.go]: Done

**Comment by Jiaqicao257** [comment]
@grac3gao yes, it's manually tested on A100 40GB with partition size 1g.5gb and on A100 80GB and H100 80GB with partition size 1g.10gb

**Review by grac3gao — APPROVED** [review]

**Pull request merged**